dash/include/dash/halo/Halo.h

```diff
@@ -0,0 +1,1104 @@
+        bnd_elem_extents[d_tmp] -=
+          halo_extents_max[d_tmp].first + halo_extents_max[d_tmp].second;
+      }
+
+      _view_inner.resize_dim(
+        d, view_offset + halo_extents_max[d].first,
+        view_extent - halo_extents_max[d].first - halo_extents_max[d].second );
+      if( cycle_spec[d] == Cycle::NONE ) {
+        auto safe_offset = view_offset;
+        auto safe_extent = view_extent;
+        if( view_offset < halo_extents_max[d].first ) {
+          safe_offset = halo_extents_max[d].first;
+          safe_extent -= halo_extents_max[d].first - view_offset;
+        } else {
+          pushBndElems( d, bnd_elem_offsets, bnd_elem_extents, halo_extents_max,
+                        cycle_spec );
+        }
+        auto check_extent =
+          view_offset + view_extent + halo_extents_max[d].second;
+        if( check_extent > _pattern.extent( d ) ) {
+          safe_extent -= check_extent - _pattern.extent( d );
+        } else {
+          bnd_elem_offsets[d] += view_extent - halo_extents_max[d].first;
+          bnd_elem_extents[d] = halo_extents_max[d].second;
+          pushBndElems( d, bnd_elem_offsets, bnd_elem_extents, halo_extents_max,
+                        cycle_spec );
+        }
+        _view_safe.resize_dim( d, safe_offset, safe_extent );
+      } else {
+        pushBndElems( d, bnd_elem_offsets, bnd_elem_extents, halo_extents_max,
+                      cycle_spec );
+        bnd_elem_offsets[d] += view_extent - halo_extents_max[d].first;
+        bnd_elem_extents[d] = halo_extents_max[d].second;
+        pushBndElems( d, bnd_elem_offsets, bnd_elem_extents, halo_extents_max,
+                      cycle_spec );
+      }
+    }
+  }
+
+  HaloBlock() = delete;
+
+  /**
+   * Copy constructor.
+   */
+  HaloBlock( const SelfT& other ) = default;
+
+  /**
+   * Assignment operator.
+   */
+  SelfT& operator=( const SelfT& other ) = default;
+
+  static constexpr dim_t ndim() { return NumDimensions; }
+
+  /**
+   * The pattern instance that created the encapsulated block.
+   */
+  const PatternT& pattern() const { return _pattern; }
+
+  const GlobMemT& globmem() const { return _globmem; }
+
+  const region_t* halo_region( const region_index_t index ) const {
+    return _halo_reg_mapping[index];
+  }
+
+  const regions_t& halo_regions() const { return _halo_regions; }
+
+  const region_t* boundary_region( const region_index_t index ) const {
+    return _boundary_reg_mapping[index];
+  }
+
+  const regions_t& boundary_regions() const { return _boundary_regions; }
+
+  // const halo_regs_t& halo_regions() const { return _halo_regions; }
+
+  const ViewSpecT& view() const { return _view; }
+
+  const ViewSpecT& view_safe() const { return _view_safe; }
+
+  const ViewSpecT& view_inner() const { return _view_inner; }
+
+  const std::vector<ViewSpecT>& boundary_elements() const {
+    return _boundary_elements;
+  }
+
+  size_type halo_size() const {
```

Comment:
it seems that the size can be computed once during setup and cached.

Reply:
true